pkg/sql/sessioninit/cache.go

```diff
@@ -146,53 +144,43 @@ func (a *Cache) GetAuthInfo(
 		// trying to cache anything.
 		if usersTableDesc.IsUncommittedVersion() ||
 			roleOptionsTableDesc.IsUncommittedVersion() {
-			aInfo, err = readFromSystemTables(
-				ctx,
-				txn,
-				ie,
-				username,
-			)
+			aInfo, err = readFromSystemTables(ctx, txn, ie, username)
 			return err
 		}
 		usersTableVersion := usersTableDesc.GetVersion()
 		roleOptionsTableVersion := roleOptionsTableDesc.GetVersion()
 
-		// We loop in case the table version changes while looking up
-		// password or role options.
-		for {
-			// Check version and maybe clear cache while holding the mutex.
-			var found bool
-			aInfo, found = a.readAuthInfoFromCache(ctx, usersTableVersion, roleOptionsTableVersion, username)
-
-			if found {
-				return nil
-			}
+		// Check version and maybe clear cache while holding the mutex.
+		var found bool
+		aInfo, found = a.readAuthInfoFromCache(ctx, usersTableVersion, roleOptionsTableVersion, username)
 
-			// Lookup the data outside the lock, with at most one request in-flight
-			// for each user. The user and role_options table versions are also part
-			// of the request key so that we don't read data from an old version
-			// of either table.
-			val, err := a.loadCacheValue(
-				ctx, fmt.Sprintf("authinfo-%s-%d-%d", username.Normalized(), usersTableVersion, roleOptionsTableVersion),
-				func(loadCtx context.Context) (interface{}, error) {
-					return readFromSystemTables(loadCtx, txn, ie, username)
-				})
-			if err != nil {
-				return err
-			}
-			aInfo = val.(AuthInfo)
+		if found {
+			return nil
+		}
 
-			finishedLoop := a.writeAuthInfoBackToCache(
-				ctx,
-				&usersTableVersion,
-				&roleOptionsTableVersion,
-				aInfo,
-				username,
-			)
-			if finishedLoop {
-				return nil
-			}
+		// Lookup the data outside the lock. There will be at most one
+		// request in-flight for each user. The user and role_options table
+		// versions are also part of the request key so that we don't read data
+		// from an old version of either table.
+		val, err := a.loadCacheValue(
+			ctx, fmt.Sprintf("authinfo-%s-%d-%d", username.Normalized(), usersTableVersion, roleOptionsTableVersion),
+			func(loadCtx context.Context) (interface{}, error) {
+				return readFromSystemTables(loadCtx, txn, ie, username)
+			})
+		if err != nil {
+			return err
 		}
+		aInfo = val.(AuthInfo)
+
+		// Write data back to the cache if the table version hasn't changed.
+		a.maybeWriteAuthInfoBackToCache(
```

Comment:
remind me why we don't write to the cache whilst in the lock? doesn't that mean until the cache is populated all entries fail during the "thundering hurd"?

Reply:
also should the singleflight should double check the cache again just in case it is populated?

sorry i should've inquired earlier

Reply:
> remind me why we don't write to the cache whilst in the lock/singleflight? doesn't that mean until the cache is populated all entries fail during the "thundering hurd"?

we do write to the cache while holding the lock. `maybeWriteAuthInfoBackToCache` acquires the lock

> doesn't that mean until the cache is populated all entries fail during the "thundering hurd"?

what do you mean by all entries fail? anyway, hopefully your concern is resolved since we do take the lock when writing to the cache

> also should the singleflight should double check the cache again just in case it is populated?

this doesn't seem like it would save much I/O. if there already is a request in flight, we'll just delegate to that request. i guess there is a race condition (not a correctness one) if:
- T1 has a cache miss
- T1 starts the request and completes it
- before T1 writes back to the cache, T2 also gets a cache miss
- T2 starts a new request, which will retrieve the same data that T1 just did
- both T1 and T2 write the same data into the cache

however, even if we make the singleflight check the cache first, the same exact race condition is still possible.

i think the thing to fix here is to change `maybeWriteAuthInfoBackToCache` so that it doesn't grow the memory account if a different goroutine already wrote to the cache for this key.

Reply:
i wouldn't call it a race condition so much as "there may be a lot of simultaneous cache misses", causing all these requests to be backed up and send queries one at a time to resolve during auth.

i'm not going to push hard on it but i think we should probably fix that.

Reply:
thinking about this more, i'm wondering why we don't use a `RWMutex`, which mostly achieves the singleflight:

```
func() {
l.RLock()
defer l.RUnlock()

if tableVersion != version {
  needsRefresh = true
}

aInfo, ok = ....
}()

if ok {
   return aInfo
}

l.Lock()
defer l.Unlock()

// redo checks
// if version < tableVersion singleflight a check
// otherwise, read checks and populate cache
```

Reply:
> causing all these requests to be backed up and send queries one at a time to resolve during auth.

ah i think that's the misunderstanding. the requests don't get backed up. the singleflight request happens outside of the lock. the way singleflight works is that if T2 tries to make a request for a requestKey, and T1 has already started that request but is waiting to hear back, then T2 will not make *any* request at all. it will just wait for T1 to get the answer, and use that same response.

based on your questions, it sounds like you were thinking that singleflight causes all the requests to queue up, then they all happen one at a time. in fact it's quite the opposite. we want all the lookups to the system tables happen  concurrently, which it why they are outside of the mutex. all singleflight does is make it so these concurrent requests don't duplicate the same lookup.

actually your example has worse concurrency -- it's holding the write lock _during_ the lookup from system tables. this means it could hold a write lock for a very expensive call (up to 1 second).

Reply:
oh i see. does that make it a little iffy if there are different txns? should it be nil?

Reply:
> oh i see. does that make it a little iffy if there are different txns? should it be nil?

no, since we use the table versions in the request key. so T2 will only read from T1 if the two transactions are reading from the same version of the table.

Reply:
ah! witchcraft